### DISCUSSION TITLE
Private View Abstraction for Collection / Items

### DISCUSSION THREAD

**Discussion by ernst86913** [directus/directus#4972] [Feature Requests]
Hi - 
in order to have a cleaner and more maintainable APP code I would suggest the following changes / re-factoring:

- Keep current private-view component as top level... but modules should normally not use it directly (only if really custom)
- Create a private-view-item component which is used by ALL item pages of modules... and which provides item specific functionality (header and action buttons, form options, sidebar item specifics incl. quick info, item settings, comments,....)
- Create a private-view-collection component which is used by ALL collection pages of modules... and which provides collection specific functionality (search, filters, settings,....)

Further I would suggest that the system modules / pages make use of those components as well as those are currently separate and lack integration with the rest.

Any thoughts?

KR
Roman

**Comment by rijkvanzanten** [comment]
> Keep current private-view component as top level... but modules should normally not use it directly (only if really custom)

Care to expand on this?

> Create a private-view-item component which is used by ALL item pages of modules... and which provides item specific functionality (header and action buttons, form options, sidebar item specifics incl. quick info, item settings, comments,....)
> Create a private-view-collection component which is used by ALL collection pages of modules... and which provides collection specific functionality (search, filters, settings,....)

I'm not really following what `private-view-item` and `private-view-collection` would be in this case 🤔

**Reply by ernst86913** [reply]
The main questions I am asking is.... how can we....

a) Allow for a standardized COLLECTIONS and ITEMS view for all Collection components
b) But - allow for configurable adjustments of those based on COLLECTION-NAME  (e.g. Header, Content, SideBar,...)
c) Without having to re-write it as a full custom module.

Examples:
- For collections type "invoice" we want to add some statistics to the sidebar... and on top to the standard collections buttons there should be 2 additional buttons in the header
- For collections type "customer" we want to add a header in the content like you already did with USERS

How can this be achieved in a configurable way without re-creating the full already existing collections module, which would result in code duplication and bad maintainability.... 

Hope it makes more sense now :-)

**Reply by rijkvanzanten** [reply]
Ahh yeah that makes total sense. That's actually how it used to work in v7/v8, but that turned into a huge mess.. The amount of additional overrides and functionality between the different "versions" of the page meant it was actually easier to duplicate the minimal boilerplate, and re-use shared bits through composables (that being said, a little more could be shared currently). I'd rather not go back to having that setup where it's trying to reuse too much which then limits us in terms of what sorts of overrides can be done in the individual instances

**Reply by ernst86913** [reply]
Fully understand your view.

From your perspective - what would be the best way to dynamically load addtl. elements based on Collection type? Any recommendations?

**Comment by jaads** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
